### PR TITLE
Issue 1181

### LINKED ISSUE #1181: Dates in nextUpdateTime are incorrect
Looks like the dates aren't accurate, which is likely preventing the DNT policy from being rechecked for users and allowing traffic through to newly compliant domains.

### PR DESCRIPTION
**Pull request by cooperq**
fixes #1181

### CODE REVIEW

File: src/migrations.js [hunk 1]
```diff
@@ -54,12 +54,23 @@ exports.Migrations= {
     for(var domain in action_map.getItemClones()){
       if(badger.storage.getNextUpdateForDomain(domain) === 0){
         // Recheck at a random time in the next week
-        var recheckTime = utils.getRandom(Date.now(), Date.now() + (utils.oneDay() * 7));
+        var recheckTime = utils.getRandom(utils.nDaysFromNow(7), utils.nDaysFromNow(14));
```
Comment (ghostwords) [inline src/migrations.js]: The original logic seems fine (sometime between now and seven days from now) ... We're using the new wrapper functions but with a changed range.

**Comment by ghostwords** [comment]
New utils method looks good. In PB's background page context:
```javascript
console.log("now:", new Date());
var set = _.times(1000, () => {
  return utils.getRandom(utils.oneDayFromNow(), utils.nDaysFromNow(7));
});
console.log("min:", new Date(_.min(set)));
console.log("max:", new Date(_.max(set)));
```
Sample output:
```
now: Mon Feb 06 2017 14:36:57 GMT-0500 (EST)
min: Tue Feb 07 2017 14:47:25 GMT-0500 (EST)
max: Mon Feb 13 2017 14:26:11 GMT-0500 (EST)
```

**Comment by ghostwords** [comment]
Earlier discussion from when our Firefox addon data migration caused browser lockups: https://github.com/EFForg/privacybadger/pull/1082#issuecomment-267431397

**Comment by ghostwords** [comment]
My bad, debouncing doesn't really apply, what we actually want is rate limiting of requests. Build a queue of DNT requests, send them at regular intervals. New ones get added to the queue.

To avoid making multiple requests for the same domain, we should check that the domain isn't already in the queue and that we are not in the middle of a check for that domain, in addition to checking `nextUpdateTime`. Maybe no need for new flag to indicate that we're in the middle of checking: just update `nextUpdateTime` before making the XHR instead of after?

**Comment by cooperq** [comment]
fixes #809

**Comment by ghostwords** [comment]
Just came across another couple of DNT-related bugs. Periodic policy and hash checking was broken; `setInterval` binds to global scope and so `this` inside the callback ends up referencing the window object instead of Badger. Fixed in db6c5e09fdc4a17947b497e6f6920c0095d5b1db.

**Comment by ghostwords** [comment]
Ditto for yellowlist updating: https://github.com/EFForg/privacybadger/pull/1182/commits/3dec4b945a532198b580be46af843bde29d6a2fb

**Comment by ghostwords** [comment]
@cooperq please review. Looks like GitHub won't let me assign this to you for review officially, probably since it's your pull request.

**Comment by cooperq** [comment]
This seems to reintroduce some intermittent test failures. Any thoughts on this alexei?

**Comment by cooperq** [comment]
specifically it looks like there is now a race condition in the "checks do not track policy" test on pbtest.org

**Comment by ghostwords** [comment]
I merged in master late; there is only one commit afterwards. It looks like the pbtest.org suite is retried many times now but it still managed to fail for my latest commit (looks like "Does privacy badger unblock 3rd parties that promise to honor Do Not Track?" failed, which seems relevant ...). Is this an actual problem or are pbtest.org just really flaky still (Why are they still so flaky? When did they get so flaky? I feel like they weren't always this bad.)? Do tests pass for you locally? Mine do, at least in Chrome.

**Comment by cooperq** [comment]
Well I think maybe we have just introduced a race condition by implementing the queue and debouncing features in this case. (I think we had previously introduced a race condition in a different test).

**Comment by ghostwords** [comment]
@alexristich Internet being down results in each check failing quickly (with ERR_INTERNET_DISCONNECTED XMLHttpRequest errors). Similarly to a restart with unprocessed items in the queue, such domains are effectively skipped and will once again be eligible for checking some time between one and seven days from now. I think this is fine.

**Comment by ghostwords** [comment]
@alexristich Regarding `this` working correctly in the rateLimit callback, I think the reason is that in JavaScript the value of `this` gets determined (implicitly) at call time, and when we call `badger.checkPrivacyBadgerPolicy`, our context is `badger`, which came from a constructor call and has the appropriate value of `this`.

The fact that `badger.checkPrivacyBadgerPolicy` is the result of a function call that took the original definition as an argument does not have any bearing on the value of `this` inside the original definition since the function call didn't explicitly rebind `this`.

Here is a reduced scenario, which returns `"bar"` (and not `undefined`).
```javascript
function rateLimit(fn) {
  // Note how this prints the window and not the Badger object
  // because that's what `this` is when rateLimit is called.
  console.log(this);
  return function () { return fn(); };
}
function Badger() { this.foo = "bar"; }
Badger.prototype = { checkPrivacyBadgerPolicy: rateLimit(function () { return this.foo; }), };
var badger = new Badger();
badger.checkPrivacyBadgerPolicy();
```

**Comment by ghostwords** [comment]
Edit: We need to do `fn.apply(this)` and not just `fn()` inside `rateLimit` since although `this` is correct (it's a Badger) inside `rateLimit`, calling `fn` without any explicit binding will rebind `this` inside `fn` to window as per JavaScript's call time context binding.

**Comment by ghostwords** [comment]
See also: [arrow functions](https://rainsoft.io/when-not-to-use-arrow-functions-in-javascript/), [strict mode](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Strict_mode/Transitioning_to_strict_mode#this_in_function_calls).

**Comment by ghostwords** [comment]
I noticed a new exception in the background page console, need to investigate:
```
Uncaught TypeError: Cannot set property 'nextUpdateTime' of null
    at BadgerPen.touchDNTRecheckTime (storage.js:114)
    at Badger.checkForDNTPolicy (background.js:350)
    at heuristicblocking.js:160
```

**Comment by ghostwords** [comment]
The exception is a consequence of moving up the DNT recheck timestamp update to happen first, instead of upon DNT check XHR's completion, in a3a5cda0f9fbeb43adea7f4171efe18d4126f416. Investigating ...

**Comment by ghostwords** [comment]
@cooperq Tests pass in Chrome, fail in Firefox, but also fail on master in the same way.

**Comment by cooperq** [comment]
LGTM

**Comment by ghostwords** [comment]
Opened #1245 and #1247 to fix failing Firefox tests.

**Pull request merged**